### ISSUE TITLE
Safari Support

### ISSUE THREAD

**Issue by BenMcGarry** [honestbleeps/Reddit-Enhancement-Suite#5238]
So Apple just announced the adoption of WebExtensions API in Big Sur. Using this issue as a track for testing and other notes. No way a confirmation we will be supporting Safari.

**Comment by creesch** [comment]
See also https://www.reddit.com/r/Enhancement/comments/hdxq52/so_will_you_bring_back_safari_support_now/

**Comment by BenMcGarry** [comment]
Command line is used to convert project to xcode compatible to be submitted to the store for apple review and publish. Dev fee is required. https://developer.apple.com/documentation/safariservices/safari_web_extensions/converting_a_web_extension_for_safari

https://developer.apple.com/documentation/safariservices/safari_web_extensions/assessing_your_safari_web_extension_s_browser_compatibility

We already potentially have manifest conflicts (unlimitedStorage is 10MB??????)

**Comment by honestbleeps** [comment]
not actually tested, just based on my reading of these docs, a few things that may cause issues for RES features in Safari... Docs: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Browser_support_for_JavaScript_APIs

Issues:
- downloads API support (lack thereof) - no naming images after title on download, I think?
- events support - not sure if this immediately impacts RES or not without reviewing the code, but noting it because it's severely lacking
- history support - lack of history support will mean no marking links "read" in Safari
- incognito support - lack of incognito support means possibly not being able to enable RES in private browsing mode
- identity support - I'm wild guessing here as I didn't implement the google drive / other backup support, but might this prevent saving settings to cloud storage services?
- pageAction - no hide support means subreddit CSS toggle icon (toolbar button) might appear even when not browsing reddit


By no means do I feel this is exhaustive, it's just a first pass at things that stick out that I know we use in RES that might be issues for certain features.

**Comment by BenMcGarry** [comment]
I should have a full idea of where we stand shortly. Grabbing all the console errors I can to see what's broken. But a lot of the core APIs we need just aren't simply there in 14.

**Comment by larsjohnsen** [comment]
The missing APIs do not seem to be too big an obstacle:

> identity

Chrome does not currently use this API due to some cookie problem.

>    downloads (Not supported at all)

Only used for showImages, not a big loss - easy workaround. Old Edge also lacked this.

>    history (Not supported at all)

In addition to what honestbleeps mentions, Safari will also lack the the `isRead` filter. Though this also didn't work in the old Edge.

>    unlimitedStorage (Capped at 10MB)

Likely very few that have that much in storage. We could add a module which forces pruning of caches when this limit is hit.

>    i18n (Limited support, missing API we rely on)

We use `requestIdleCallback` which is easily polyfillable (this was also the case for Firefox a few versions ago).

**Comment by BenMcGarry** [comment]
We will end up wasting time trying to get polyfills in when a limited amount of us can actually test this. My opinion right now is to hold off until they mature the API landscape. It was a PITA adding in polyfills for Edge and removing them all and unifying the code was a big project. Adding them back feels dirty.

**Comment by BenMcGarry** [comment]
@larsjohnsen Realistically, how much polyfill do you think would be required? Like would it be possible to polyfill with a separate js loaded in to prevent heavy code change?

**Comment by ShotSkydiver** [comment]
So: exciting news on the Safari extensions front! macOS Big Sur beta 2 was just released yesterday, and as I was reading through the release notes, I noticed this line: 

> Support for Web Extensions is now available. Existing Chrome and Firefox extensions can be converted for Safari using xcrun safari-web-extension-converter and distributed through the App Store for use in Safari 14. (55707949)

Needless to say, I immediately got excited and tried it out for myself, using a bunch of different chrome extensions as tests, and long story short, I've gotten to a point where I now have RES running as an extension in Safari 14! (v5.2.2; the latest release had some issues but are most likely easily fixable). So far I've tested this new conversion tool out with Imagus, Lovely Forks, and Cloud to Butt as well, and all work (mostly) perfectly! This could be huge for RES on Safari. I'm going to spend more time playing around with trying to get the latest version to work, will update if I get it working!

<img width="874" alt="Screen Shot 2020-07-08 at 4 18 57 AM" src="https://user-images.githubusercontent.com/2718920/86907665-a446f400-c0d2-11ea-90bf-477c064bbbbd.png">

<img width="2307" alt="Screen Shot 2020-07-08 at 4 19 32 AM" src="https://user-images.githubusercontent.com/2718920/86907669-a4df8a80-c0d2-11ea-8b23-934670b6e9f4.png">

**Comment by ShotSkydiver** [comment]
> > i18n (Limited support, missing API we rely on)
> 
> We use `requestIdleCallback` which is easily polyfillable (this was also the case for Firefox a few versions ago).

FYI: In Safari 14 at least, there's an optional Experimental Feature toggle to enable `requestIdleCallback`. 

<img width="905" alt="Screen Shot 2020-07-08 at 4 14 03 AM" src="https://user-images.githubusercontent.com/2718920/86907781-cf314800-c0d2-11ea-9bde-bf3d5d69d650.png">

**Comment by BenMcGarry** [comment]
The version of RES you are testing is **extremely** old. v5.2.2 was released in December 2016 and is no longer a real representation of the RES code base as its gone through extensive changes. That release was actually prior to RES adopting webextensions.

**Comment by unbeatable-101** [comment]
> Extensions must be delivered via the App Store. Meaning the $99 is still required and we cannot deliver another way. Still subject to the review process.

Then how does adguard offer a direct download from their site?
https://adguard.com/en/adguard-mac/overview.html

**Comment by honestbleeps** [comment]
@unbeatable-101 there's far more to this than the release process, but I'll try and summarize as succinctly as possible:

1) there was once tons of Safari-unique code in RES because of its various quirks, differing API functionality, etc. That has since been removed (now years ago) since support was dropped and it was extra work to maintain and test separate code on another browser for every release.

2) The new support Safari has for chrome extension APIs is actually VERY limited. Many features of RES would not work on Safari even if we supported it, because its API support is fairly limited. This may become less true over time, but it is true for now.

3) RES is used by millions of users, and therefore being able to push automatic updates to users is critical to us in order to minimize support requests, increase users' satisfaction (through prompt bug fixes, etc), and also to ensure that any security updates are pushed out in a timely manner.  Last I checked, self-hosting, as Adguard does, does not allow for automatic updating. 

4) Beyond the above, it's also significant extra work to build for Safari, as xcode is required, etc. First we must own macs just to build/test it - not everyone can afford the expensive hardware.  Next, a certain amount of extra work is required for every build, and for every test -- so even if we have macs (not all contributors do), it's significant extra overhead to support Safari.

We will continue to watch as Safari's support hopefully improves, and we *may* revisit supporting it at some point - but at this point, attempting to convince us we're wrong not to support it isn't really helpful. What would be helpful is if someone wants to dedicate time to contributing to its support. I'd consider paying the fee and putting it into the extension gallery IF we can get adequate support not only from Apple and Safari's APIs, but from some number of developers willing to continue supporting it over time - because every time we release new features, bug fixes, etc, they have to be tested across yet one more browser - adding overhead to every release.

TL;DR: it's just not as simple as "you guys are being [stubborn|objectionable|lazy|contrarian] for not supporting Safari".

**Comment by unbeatable-101** [comment]
I know all that, I’m just talking about the App Store problem, why is it required?

**Comment by unbeatable-101** [comment]
I knew you guys weren’t just being lazy, I was just wondering about the adguard part. Thank you for the answer

**Comment by BenMcGarry** [comment]
It looks like AdGuard for Mac is an actual OSX application, which apple allow you to host and provide download. However for Safari extensions Apple strictly require them to only be available via the store.

**Comment by dvcrn** [comment]
Refined Github went the route of building an .app and loading it by enabling "Allow Unsigned Extensions" - https://github.com/sindresorhus/refined-github/issues/14

I think RES could probably do something similar

**Comment by stephensobel-toast** [comment]
according to the latest comment on that thread, users will have to build it locally, so it's still a pretty unfriendly method... See here: https://github.com/sindresorhus/refined-github/issues/14#issuecomment-716076857

and there's still the matters of:

- Safari's API doesn't support all webextensions APIs we'd want for RES 
- Safari's unique bugs we'd have to seek out and put fixes back in for - it was always the browser that required the most custom, browser-specific code to avoid issues

The workload still outweighs the benefits at this point, probably.

**Comment by honestbleeps** [comment]
according to the latest comment on that thread, users will have to build it locally, so it's still a pretty unfriendly method... See here: sindresorhus/refined-github#14 (comment)

and there's still the matters of:

- Safari's API doesn't support all webextensions APIs we'd want for RES
- Safari's unique bugs we'd have to seek out and put fixes back in for - it was always the browser that required the most custom, browser-specific code to avoid issues

The workload still outweighs the benefits at this point, probably.

**Comment by dvcrn** [comment]
I'd love to help porting this but I'm not so familiar with this repository. Do you have a summary by chance which APIs aren't supported that RES needs that Safari doesn't support yet?

**Comment by KZeni** [comment]
@honestbleeps Maybe https://redditenhancementsuite.com/contribute/ could be updated to have an option where the donation is marked specifically towards helping support Mac & Safari development (having an active Apple developer account subscription, having necessary hardware [be it an M1 Mac Mini or a refurbished Mac to keep costs at a minimum if nobody with dev involvement has access to a Mac yet], added development time, etc.) as missing out on this outright is something some people might be willing to chip in on to help get it there if the dev team doesn't naturally expand to include a Mac dev. Could also promote awareness that you know people want it & could bring in untapped support.

**Comment by fuzzy76** [comment]
Has anyone taken a look on how this has changed in the 5 years that have passed?

**Comment by BeansIsFat** [comment]
@fuzzy76 I have converted the Firefox extension using `xcrun safari-web-extension-converter` and applied a few fixes to function calls that were causing issues in Safari. It seems to be working for me, so my next step is to write a script to apply the patches to a fresh download of the extension.

If you're interested, I can share the script once I'm done. You'll need to compile the project with Xcode and ideally sign it with a personal developer certificate so you can run it on your own machine without having to enable unsigned extensions every time you restart Safari.

**Comment by fuzzy76** [comment]
Thank you for your work, and for showing that Safari is a viable target for RES. However, I prefer to wait for a packaged final version. :)

**Comment by KZeni** [comment]
@fuzzy76 True, it's awesome to see steps put in place to have a Safari version of the extension made available without too much maintenance (converted & patched and then having it automate that whenever a new version becomes available), but I'm hoping that will also then see it given proper code signing & distribution via the mac/ios/etc. app store to get a fully streamlined experience for the user.
